webapp/channels/src/utils/popouts/browser_popouts.ts

```diff
@@ -0,0 +1,168 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import {useEffect} from 'react';
+import {useHistory} from 'react-router-dom';
+
+import {getHistory} from 'utils/browser_history';
+import {isDesktopApp} from 'utils/user_agent';
+
+const POPOUT_WIDTH = 800;
+const POPOUT_HEIGHT_OFFSET = 100;
+const CLOSE_POLL_INTERVAL = 200;
+const CLOSE_POLL_MAX_TRIES = 10;
+
+export function useBrowserPopout() {
+    const history = useHistory();
+    useEffect(() => {
+        if (!isDesktopApp()) {
+            const unblockHistory = history.block((blockState) => {
+                window.opener.postMessage({
+                    channel: '_navigate',
+                    args: [blockState.pathname],
+                }, window.location.origin);
+                return false;
+            });
+            const closeListener = () => {
+                window.opener.postMessage({
+                    channel: '_close',
+                    args: [],
+                }, window.location.origin);
+            };
+
+            window.addEventListener('beforeunload', closeListener);
+            return () => {
+                unblockHistory();
+                window.removeEventListener('beforeunload', closeListener);
+            };
+        }
+
+        return () => {};
+    }, [history]);
+}
+
+export class BrowserPopouts {
+    private popoutWindows: Map<Window, string>;
+    private listeners: Map<string, Set<(channel: string, ...args: any[]) => void>>;
+    private closeListeners: Map<string, Set<() => void>>;
+
+    constructor() {
+        this.popoutWindows = new Map();
+        this.listeners = new Map();
+        this.closeListeners = new Map();
+
+        window.addEventListener('message', (event) => {
+            const popout = event.source as Window;
+            const target = this.popoutWindows.get(popout);
+
+            if (event.origin !== window.location.origin) {
+                return;
+            }
+
+            if (!target) {
+                return;
+            }
+
+            switch (event.data.channel) {
+            case '_navigate':
+                window.focus();
```

Comment:
Are there any issues if this fails due to the browser deciding to not do this? 0/5 since there's probably nothing we can do in that case